### PR TITLE
8213445: jcmd VM.symboltable and VM.stringtable -verbose output contains no shared symbols or strings

### PR DESCRIPTION
**Pull request by calvinccheung**
Please review this change which will include shared symbols and strings when using the commands `jcmd VMsymboltable -verbose` and `jcmd VM.stringtable -verbose`, respectively.

This change also prints the shared symbol and shared string tables statistics if the `-verbose` option is not specified.

An example output of shared symbols:

```
48084:
VERSION: 1.1
25 2: [Ljava/util/jar/Manifest;
13 65535: jdk/jfr/Event
26 2: [Ljava/lang/AutoCloseable;
5 1: 'app'
10 1: 'platform'
24 65535: jdk/internal/event/Event
---------------
Shared symbols:
---------------
0 65535:
10 65535: toIndex =
3 65535: sbc
14 65535: package.access
...
---------------
Dynamic shared symbols:
---------------
19 65535: findResourceInPatch
71 65535: (Ljava/util/List<Ljdk/internal/module/ModulePatcher$ResourceFinder;>;)V
8 65535: overflow
49 65535: sun/net/www/protocol/jrt/JavaRuntimeURLConnection
24 65535: com/sun/tools/javac/Main
29 65535: com/sun/tools/javac/main/Main
```

Note that the `Dynamic shared symbols` section only shows up if the `jcmd` is used on a running process which is using a dynamic archive.

An example output of symbol table statistics:

```
48494:
SymbolTable statistics:
Number of buckets       :     32768 =    262144 bytes, each 8
Number of entries       :         6 =        96 bytes, each 16
Number of literals      :         6 =       152 bytes, avg  25.000
Total footprint         :           =    262392 bytes
Average bucket size     :     0.000
Variance of bucket size :     0.000
Std. dev. of bucket size:     0.014
Maximum bucket size     :         1
Shared Symbol Table statistics:
Number of buckets       :     10332
Number of entries       :     39893
Maximum bucket size     :        13
Dynamic Shared Symbol Table statistics:
Number of buckets       :        51
Number of entries       :        42
Maximum bucket size     :         3
```

Note that the `Dynamic Shared Symbol Table statistics` section only shows up if the `jcmd` is used on a running process which is using a dynamic archive.

Passed CI tiers 1 and 2 testing.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8213445](https://bugs.openjdk.java.net/browse/JDK-8213445): jcmd VM.symboltable and VM.stringtable -verbose output contains no shared symbols or strings

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 5f5a5bbb5e70efbc8257bad5fb8af84e5c243a6c
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8234/head:pull/8234` \
`$ git checkout pull/8234`

Update a local copy of the PR: \
`$ git checkout pull/8234` \
`$ git pull https://git.openjdk.java.net/jdk pull/8234/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8234`

View PR using the GUI difftool: \
`$ git pr show -t 8234`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8234.diff">https://git.openjdk.java.net/jdk/pull/8234.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
  * @requires vm.flagless
  * @library /test/lib /test/hotspot/jtreg/runtime/cds/appcds
  * @modules jdk.jcmd/sun.tools.common:+open
+ * @compile ../../../../../../lib/jdk/test/lib/apps/LingeredApp.java
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java:33]: Is this change necessary? It seems unrelated to this PR.

File: src/hotspot/share/classfile/symbolTable.cpp [hunk 2]
```diff
@@ -592,6 +612,20 @@ void SymbolTable::dump(outputStream* st, bool verbose) {
...
     }
+    if (!_shared_table.empty()) {
+      st->print_cr("---------------");
+      st->print_cr("Shared symbols:");
+      st->print_cr("---------------");
+      DumpSharedSymbol dss(st);
+      _shared_table.iterate(&dss);
+    }
+    if (!_dynamic_shared_table.empty()) {
+      st->print_cr("---------------");
```
Comment (iklam) [inline src/hotspot/share/classfile/symbolTable.cpp:623]: I think it's better to prefix the comments with `#`. That way it will be easier to include the string/symbol output into -XX: SharedArchiveConfigFile.

**Review by iklam — APPROVED** [review]
LGTM. Just a few small nits.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java (hunk 1 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java:33]: Yes. Otherwise, I saw the following failures in the tests under cds/appcds/jcmd

```
<prefix path>/testoutput/test-support/jtreg_open_test_hotspot_jtreg_hotspot_tier2_runtime/classes/2/runtime/cds/appcds/jcmd/JCmdTestFileSafety.d/jdk/test/lib/apps/LingeredApp.class : no such file or directory
<prefix path>/testoutput/test-support/jtreg_open_test_hotspot_jtreg_hotspot_tier2_runtime/classes/2/runtime/cds/appcds/jcmd/JCmdTestFileSafety.d/jdk/test/lib/apps/LingeredApp$1.class : no such file or directory
java.lang.RuntimeException: jar operation failed
	at JarBuilder.createJar(JarBuilder.java:216)
	at JarBuilder.createSimpleJar(JarBuilder.java:180)
	at JarBuilder.build(JarBuilder.java:90)
	at JCmdTestDumpBase.buildJars(JCmdTestDumpBase.java:91)
	at JCmdTestFileSafety.test(JCmdTestFileSafety.java:75)
	at JCmdTestDumpBase.runTest(JCmdTestDumpBase.java:61)
	at JCmdTestFileSafety.main(JCmdTestFileSafety.java:145)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
	at java.base/java.lang.Thread.run(Thread.java:828)
```
Not sure why my new DynamicSharedSymbols.java seems to trigger the above failure.

File: src/hotspot/share/classfile/symbolTable.cpp (hunk 2 above)
Comment (calvinccheung) [reply src/hotspot/share/classfile/symbolTable.cpp]: Fixed.

Thanks for your review.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java (hunk 1 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java:33]: > Yes. Otherwise, I saw the following failures in the tests under cds/appcds/jcmd
> 
> ```
> <prefix path>/testoutput/test-support/jtreg_open_test_hotspot_jtreg_hotspot_tier2_runtime/classes/2/runtime/cds/appcds/jcmd/JCmdTestFileSafety.d/jdk/test/lib/apps/LingeredApp.class : no such file or directory
> <prefix path>/testoutput/test-support/jtreg_open_test_hotspot_jtreg_hotspot_tier2_runtime/classes/2/runtime/cds/appcds/jcmd/JCmdTestFileSafety.d/jdk/test/lib/apps/LingeredApp$1.class : no such file or directory
> java.lang.RuntimeException: jar operation failed
> 	at JarBuilder.createJar(JarBuilder.java:216)
> 	at JarBuilder.createSimpleJar(JarBuilder.java:180)
> 	at JarBuilder.build(JarBuilder.java:90)
> 	at JCmdTestDumpBase.buildJars(JCmdTestDumpBase.java:91)
> 	at JCmdTestFileSafety.test(JCmdTestFileSafety.java:75)
> 	at JCmdTestDumpBase.runTest(JCmdTestDumpBase.java:61)
> 	at JCmdTestFileSafety.main(JCmdTestFileSafety.java:145)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
> 	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
> 	at java.base/java.lang.Thread.run(Thread.java:828)
> ```
> 
> Not sure why my new DynamicSharedSymbols.java seems to trigger the above failure.

The problem is caused by [CODETOOLS-7902847](https://bugs.openjdk.java.net/browse/CODETOOLS-7902847)

Have you tried using `@build` instead of `@compile`? That way you don't need to specify the file location.

Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java:33]: I did try `@build` but it didn't work because the class file will be placed in a path specified by the `-Dtest.class.path` property such as `./JTwork/classes/0/test/hotspot/jtreg/runtime/cds/appcds/test-classes`. During the test run, it tries to find the class from the `-Dtest.classes` such as `./JTwork/classes/0/runtime/cds/appcds/jcmd/JCmdTestFileSafety.d` and couldn't find it.

Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java]: I've updated the tests under appcds/jcmd to using `@build` and `ClassFileInstaller.writeJar` for creating jar files.

**Review by iklam — APPROVED** [review]
New changes look good to me.

File: src/hotspot/share/classfile/stringTable.cpp [hunk 3]
```diff
@@ -650,6 +654,32 @@ size_t StringTable::verify_and_compare_entries() {
   return vcs._errors;
 }
 
+static void print_string(Thread* thr, outputStream* st, oop s) {
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/stringTable.cpp:657]: Nit: if `thr` is the current thread (has to be for the `ResourceMark`) then please call it `current`.

File: src/hotspot/share/classfile/stringTable.cpp [hunk 4]
```diff
@@ -650,6 +654,32 @@ size_t StringTable::verify_and_compare_entries() {
...
+  typeArrayOop value     = java_lang_String::value_no_keepalive(s);
+  int          length    = java_lang_String::length(s);
+  bool         is_latin1 = java_lang_String::is_latin1(s);
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/stringTable.cpp:660]: Nit: there is no need to align things this way.

**Review by dholmes-ora — APPROVED** [review]
Hi Calvin,

Seems fine. A couple of style nits.

Thanks,
David

File: src/hotspot/share/classfile/stringTable.cpp (hunk 3 above)
Comment (calvinccheung) [reply src/hotspot/share/classfile/stringTable.cpp]: Fixed.

File: src/hotspot/share/classfile/stringTable.cpp (hunk 4 above)
Comment (calvinccheung) [reply src/hotspot/share/classfile/stringTable.cpp:660]: I'll leave it this time since the code was factored out from existing code.

Thanks for your review.

**Review by dholmes-ora — APPROVED** [review]

**Comment by calvinccheung** [comment]
@iklam @dholmes-ora Thanks for the review.
/integrate
